File: client/src/lib/schemas.ts
import { z } from 'zod'
import {
  SEX_OPTIONS,
  RELATIONSHIPS,
  HOLDER_RELATIONSHIPS,
  APPOINTMENT_TYPES,
  APPOINTMENT_STATUSES,
  VISIT_TYPES,
  PATIENT_STATUSES,
} from './constants'

const optionalString = z.string().optional().or(z.literal(''))

// ======================== Patient creation (multi-step) ========================

export const basicInfoSchema = z.object({
  first_name: z.string().min(1, 'First name is required'),
  middle_name: optionalString,
  last_name: z.string().min(1, 'Last name is required'),
  date_of_birth: z.string().min(1, 'Date of birth is required'),
  sex: z.enum(SEX_OPTIONS, { message: 'Sex is required' }),
  primary_language: z.string().optional().default('English'),
})
export type BasicInfoValues = z.infer<typeof basicInfoSchema>

export const contactAddressSchema = z.object({
  email: optionalString,
  phone: optionalString,
  street: z.string().min(1, 'Street address is required'),
  city: z.string().min(1, 'City is required'),
  state: z.string().min(2, 'State is required').max(2, 'Use 2-letter abbreviation'),
  zip_code: z.string().regex(/^\d{5}$/, 'ZIP code must be 5 digits'),
})
export type ContactAddressValues = z.infer<typeof contactAddressSchema>

export const emergencyContactSchema = z.object({
  name: z.string().min(1, 'Contact name is required'),
  relationship: z.enum(RELATIONSHIPS, { message: 'Relationship is required' }),
  phone: z.string().min(1, 'Phone number is required'),
  email: optionalString,
  is_primary: z.boolean().optional().default(false),
})
export type EmergencyContactValues = z.infer<typeof emergencyContactSchema>

export const emergencyContactsSchema = z.object({
  contacts: z.array(emergencyContactSchema).min(1, 'At least one contact is required'),
})
export type EmergencyContactsValues = z.infer<typeof emergencyContactsSchema>

export const insuranceSchema = z.object({
  provider_name: z.string().min(1, 'Provider name is required'),
  policy_number: z.string().min(1, 'Policy number is required'),
  group_number: optionalString,
  holder_name: z.string().min(1, 'Holder name is required'),
  holder_relationship: z.enum(HOLDER_RELATIONSHIPS, { message: 'Relationship is required' }),
})
export type InsuranceValues = z.infer<typeof insuranceSchema>

export const medicalSchema = z.object({
  primary_diagnosis: optionalString,
  allergies: optionalString,
  current_medications: optionalString,
  additional_conditions: optionalString,
})
export type MedicalValues = z.infer<typeof medicalSchema>

export const createPatientSchema = basicInfoSchema
  .merge(contactAddressSchema)
  .extend({
    emergency_contacts: z.array(emergencyContactSchema).optional(),
    insurance: insuranceSchema.optional(),
    medical: medicalSchema.optional(),
    avatar_url: z.string().url().nullable().optional(),
  })
export type CreatePatientValues = z.infer<typeof createPatientSchema>

// ======================== Patient edit schemas ========================

export const editDemographicsSchema = z.object({
  first_name: z.string().min(1, 'First name is required').optional(),
  middle_name: optionalString,
  last_name: z.string().min(1, 'Last name is required').optional(),
  date_of_birth: z.string().min(1, 'Date of birth is required').optional(),
  sex: z.enum(SEX_OPTIONS).optional(),
  primary_language: optionalString,
  email: optionalString,
  phone: optionalString,
  street: z.string().min(1, 'Street is required').optional(),
  city: z.string().min(1, 'City is required').optional(),
  state: z.string().min(2).max(2).optional(),
  zip_code: z.string().regex(/^\d{5}$/, 'ZIP code must be 5 digits').optional(),
  avatar_url: z.string().url().nullable().optional(),
})
export type EditDemographicsValues = z.infer<typeof editDemographicsSchema>

export const editStatusSchema = z.object({
  status: z.enum(PATIENT_STATUSES, { message: 'Status is required' }),
  notes: optionalString,
  referral_source: optionalString,
})
export type EditStatusValues = z.infer<typeof editStatusSchema>

// ======================== Clinical schemas ========================

export const appointmentSchema = z.object({
  provider_name: z.string().min(1, 'Provider name is required'),
  appointment_type: z.enum(APPOINTMENT_TYPES, { message: 'Type is required' }),
  date_time: z.string().min(1, 'Date and time is required'),
  duration_minutes: z.coerce.number().int().positive().optional().default(30),
  location: optionalString,
  notes: optionalString,
  status: z.enum(APPOINTMENT_STATUSES).optional().default('Scheduled'),
})
export type AppointmentValues = z.infer<typeof appointmentSchema>

export const visitSchema = z.object({
  provider_name: z.string().min(1, 'Provider name is required'),
  visit_type: z.enum(VISIT_TYPES, { message: 'Visit type is required' }),
  visit_date: z.string().min(1, 'Visit date is required'),
  summary: optionalString,
  diagnosis: optionalString,
  follow_up_needed: z.boolean().optional().default(false),
})
export type VisitValues = z.infer<typeof visitSchema>

export const immunizationSchema = z.object({
  vaccine_name: z.string().min(1, 'Vaccine name is required'),
  date_administered: z.string().min(1, 'Date administered is required'),
  administered_by: optionalString,
  dose_number: z.coerce.number().int().positive().optional(),
  lot_number: optionalString,
  next_due_date: optionalString,
  notes: optionalString,
})
export type ImmunizationValues = z.infer<typeof immunizationSchema>

export const prescriptionSchema = z.object({
  medication_id: z.string().min(1, 'Medication is required'),
  dosage: z.string().min(1, 'Dosage is required'),
  frequency: z.string().min(1, 'Frequency is required'),
  quantity: z.coerce.number().int().positive('Quantity must be positive'),
  duration: optionalString,
  pharmacy_name: z.string().min(1, 'Pharmacy is required'),
  pharmacy_address: z.string().min(1, 'Pharmacy address is required'),
  pharmacy_lat: z.coerce.number(),
  pharmacy_lng: z.coerce.number(),
  notes: optionalString,
  save_as_preferred_pharmacy: z.boolean().optional().default(false),
})
export type PrescriptionValues = z.infer<typeof prescriptionSchema>

export const pharmacySchema = z.object({
  name: z.string().min(1, 'Pharmacy name is required'),
  address: z.string().min(1, 'Address is required'),
  phone: optionalString,
  lat: z.coerce.number().optional(),
  lng: z.coerce.number().optional(),
})
export type PharmacyValues = z.infer<typeof pharmacySchema>

// ======================== Auth schemas ========================

export const loginSchema = z.object({
  email: z.string().min(1, 'Email is required').email('Invalid email address'),
  password: z.string().min(6, 'Password must be at least 6 characters'),
})
export type LoginValues = z.infer<typeof loginSchema>

export const registerSchema = z.object({
  display_name: z.string().min(1, 'Display name is required'),
  email: z.string().min(1, 'Email is required').email('Invalid email address'),
  password: z.string().min(6, 'Password must be at least 6 characters'),
  confirm_password: z.string().min(1, 'Please confirm your password'),
}).refine(data => data.password === data.confirm_password, {
  message: 'Passwords do not match',
  path: ['confirm_password'],
})
export type RegisterValues = z.infer<typeof registerSchema>

export const profileSchema = z.object({
  display_name: z.string().min(1, 'Display name is required'),
  role: z.enum(['Staff', 'Administrator', 'Provider', 'Nurse'] as const, { message: 'Role is required' }),
  phone: optionalString,
})
export type ProfileValues = z.infer<typeof profileSchema>

export const settingsSchema = z.object({
  theme: z.enum(['light', 'dark', 'system'] as const),
  notifications_enabled: z.boolean(),
  page_size: z.coerce.number().refine(v => [10, 20, 50].includes(v)),
  date_format: z.enum(['MM/DD/YYYY', 'DD/MM/YYYY', 'YYYY-MM-DD'] as const),
})
export type SettingsValues = z.infer<typeof settingsSchema>

File: server/app/schemas.py
import datetime
from typing import Optional
from uuid import UUID

from pydantic import BaseModel, ConfigDict, EmailStr


# ============================= Request schemas =============================


# --- Patient creation (multi-step flow) ---
class EmergencyContactInput(BaseModel):
    name: str
    relationship: str
    phone: str
    email: Optional[str] = None
    is_primary: Optional[bool] = False


class InsuranceInput(BaseModel):
    provider_name: str
    policy_number: str
    group_number: Optional[str] = None
    holder_name: str
    holder_relationship: str


class MedicalInput(BaseModel):
    primary_diagnosis: Optional[str] = None
    allergies: Optional[str] = None
    current_medications: Optional[str] = None
    additional_conditions: Optional[str] = None


class CreatePatientRequest(BaseModel):
    # Step 1 - Basic Info (required)
    first_name: str
    middle_name: Optional[str] = None
    last_name: str
    date_of_birth: datetime.date
    sex: str
    primary_language: Optional[str] = "English"
    # Step 2 - Contact & Address (required)
    email: Optional[str] = None
    phone: Optional[str] = None
    street: str
    city: str
    state: str
    zip_code: str
    # Step 3 - Emergency Contacts (optional)
    emergency_contacts: Optional[list[EmergencyContactInput]] = None
    # Step 4 - Insurance (optional)
    insurance: Optional[InsuranceInput] = None
    # Step 5 - Medical Info (optional)
    medical: Optional[MedicalInput] = None


# --- Demographics edit ---
class EditDemographicsRequest(BaseModel):
    first_name: Optional[str] = None
    middle_name: Optional[str] = None
    last_name: Optional[str] = None
    date_of_birth: Optional[datetime.date] = None
    sex: Optional[str] = None
    primary_language: Optional[str] = None
    email: Optional[str] = None
    phone: Optional[str] = None
    street: Optional[str] = None
    city: Optional[str] = None
    state: Optional[str] = None
    zip_code: Optional[str] = None


# --- Status edit ---
class EditStatusRequest(BaseModel):
    status: str
    notes: Optional[str] = None
    referral_source: Optional[str] = None


# --- Contacts edit (PUT = full replacement) ---
class EditContactsRequest(BaseModel):
    contacts: list[EmergencyContactInput]


# --- Insurance edit (upsert) ---
class EditInsuranceRequest(InsuranceInput):
    pass


# --- Medical edit (upsert) ---
class EditMedicalRequest(MedicalInput):
    pass


# --- Pharmacy edit (upsert) ---
class EditPharmacyRequest(BaseModel):
    name: str
    address: str
    phone: Optional[str] = None
    lat: Optional[float] = None
    lng: Optional[float] = None


# --- Appointments ---
class CreateAppointmentRequest(BaseModel):
    provider_name: str
    appointment_type: str
    date_time: datetime.datetime
    duration_minutes: Optional[int] = 30
    location: Optional[str] = None
    notes: Optional[str] = None
    status: Optional[str] = "Scheduled"


class EditAppointmentRequest(BaseModel):
    provider_name: Optional[str] = None
    appointment_type: Optional[str] = None
    date_time: Optional[datetime.datetime] = None
    duration_minutes: Optional[int] = None
    location: Optional[str] = None
    notes: Optional[str] = None
    status: Optional[str] = None


# --- Visits ---
class CreateVisitRequest(BaseModel):
    provider_name: str
    visit_type: str
    visit_date: datetime.date
    summary: Optional[str] = None
    diagnosis: Optional[str] = None
    follow_up_needed: Optional[bool] = False


class EditVisitRequest(BaseModel):
    provider_name: Optional[str] = None
    visit_type: Optional[str] = None
    visit_date: Optional[datetime.date] = None
    summary: Optional[str] = None
    diagnosis: Optional[str] = None
    follow_up_needed: Optional[bool] = None


# --- Immunizations ---
class CreateImmunizationRequest(BaseModel):
    vaccine_name: str
    date_administered: datetime.date
    administered_by: Optional[str] = None
    dose_number: Optional[int] = None
    lot_number: Optional[str] = None
    next_due_date: Optional[datetime.date] = None
    notes: Optional[str] = None


class EditImmunizationRequest(BaseModel):
    vaccine_name: Optional[str] = None
    date_administered: Optional[datetime.date] = None
    administered_by: Optional[str] = None
    dose_number: Optional[int] = None
    lot_number: Optional[str] = None
    next_due_date: Optional[datetime.date] = None
    notes: Optional[str] = None


# --- Prescriptions ---
class CreatePrescriptionRequest(BaseModel):
    medication_id: UUID
    dosage: str
    frequency: str
    quantity: int
    duration: Optional[str] = None
    pharmacy_name: str
    pharmacy_address: str
    pharmacy_lat: float
    pharmacy_lng: float
    notes: Optional[str] = None
    save_as_preferred_pharmacy: Optional[bool] = False


# ============================ Response schemas =============================


# --- Entity responses ---
class PatientResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: UUID
    first_name: str
    middle_name: Optional[str] = None
    last_name: str
    date_of_birth: datetime.date
    sex: str
    primary_language: Optional[str] = None
    email: Optional[str] = None
    phone: Optional[str] = None
    street: str
    city: str
    state: str
    zip_code: str
    status: str
    referral_source: Optional[str] = None
    notes: Optional[str] = None
    created_at: datetime.datetime
    updated_at: datetime.datetime


class EmergencyContactResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: UUID
    patient_id: UUID
    name: str
    relationship: str
    phone: str
    email: Optional[str] = None
    is_primary: Optional[bool] = False


class InsuranceInfoResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: UUID
    patient_id: UUID
    provider_name: str
    policy_number: str
    group_number: Optional[str] = None
    holder_name: str
    holder_relationship: str


class MedicalInfoResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: UUID
    patient_id: UUID
    primary_diagnosis: Optional[str] = None
    allergies: Optional[str] = None
    current_medications: Optional[str] = None
    additional_conditions: Optional[str] = None


class PreferredPharmacyResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: UUID
    patient_id: UUID
    name: str
    address: str
    phone: Optional[str] = None
    lat: Optional[float] = None
    lng: Optional[float] = None


class AppointmentResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: UUID
    patient_id: UUID
    provider_name: str
    appointment_type: str
    date_time: datetime.datetime
    duration_minutes: Optional[int] = 30
    location: Optional[str] = None
    notes: Optional[str] = None
    status: str
    created_at: datetime.datetime


class VisitResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: UUID
    patient_id: UUID
    provider_name: str
    visit_type: str
    visit_date: datetime.date
    summary: Optional[str] = None
    diagnosis: Optional[str] = None
    follow_up_needed: Optional[bool] = False
    created_at: datetime.datetime


class ImmunizationResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: UUID
    patient_id: UUID
    vaccine_name: str
    date_administered: datetime.date
    administered_by: Optional[str] = None
    dose_number: Optional[int] = None
    lot_number: Optional[str] = None
    next_due_date: Optional[datetime.date] = None
    notes: Optional[str] = None


class MedicationResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: UUID
    name: str
    generic_name: Optional[str] = None
    description: Optional[str] = None
    common_dosages: str
    form: str
    category: str


class PrescriptionResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: UUID
    patient_id: UUID
    medication_id: UUID
    dosage: str
    frequency: str
    quantity: int
    duration: Optional[str] = None
    pharmacy_name: str
    pharmacy_address: str
    pharmacy_lat: float
    pharmacy_lng: float
    prescribed_at: datetime.datetime
    notes: Optional[str] = None
    medication: Optional[MedicationResponse] = None


class StatusHistoryResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: UUID
    patient_id: UUID
    old_status: Optional[str] = None
    new_status: str
    changed_at: datetime.datetime


# --- Composite / list responses ---
class PatientDetailResponse(PatientResponse):
    emergency_contacts: list[EmergencyContactResponse] = []
    insurance_info: Optional[InsuranceInfoResponse] = None
    medical_info: Optional[MedicalInfoResponse] = None
    preferred_pharmacy: Optional[PreferredPharmacyResponse] = None
    appointments: list[AppointmentResponse] = []
    visits: list[VisitResponse] = []
    immunizations: list[ImmunizationResponse] = []
    prescriptions: list[PrescriptionResponse] = []
    status_history: list[StatusHistoryResponse] = []


class PatientListItem(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: UUID
    first_name: str
    last_name: str
    date_of_birth: datetime.date
    status: str
    city: str
    state: str
    phone: Optional[str] = None
    created_at: datetime.datetime


class PatientListResponse(BaseModel):
    items: list[PatientListItem]
    total: int
    page: int
    page_size: int
    total_pages: int


# --- Dashboard stats ---
class StatusCount(BaseModel):
    status: str
    count: int


class StatsOverviewResponse(BaseModel):
    status_counts: list[StatusCount]
    total_patients: int


class TrendDataPoint(BaseModel):
    period: str
    inquiry: int = 0
    onboarding: int = 0
    active: int = 0
    churned: int = 0


class StatsTrendsResponse(BaseModel):
    trends: list[TrendDataPoint]


# --- Recent Appointments (dashboard) ---
class RecentAppointmentItem(BaseModel):
    model_config = ConfigDict(from_attributes=True)
    id: UUID
    patient_id: UUID
    patient_name: str
    provider_name: str
    appointment_type: str
    date_time: datetime.datetime
    duration_minutes: int | None
    status: str


class RecentAppointmentsResponse(BaseModel):
    upcoming: list[RecentAppointmentItem]
    total_scheduled: int
    total_completed: int
    total_cancelled: int
    total_no_show: int


# --- Care Gaps ---
class CareGapItem(BaseModel):
    model_config = ConfigDict(from_attributes=True)
    patient_id: UUID
    patient_name: str
    visit_date: datetime.date
    diagnosis: str | None
    days_since_visit: int


class CareGapsResponse(BaseModel):
    items: list[CareGapItem]
    total_count: int


# --- New Patients Trend ---
class NewPatientsDataPoint(BaseModel):
    period: str
    count: int


class NewPatientsTrendResponse(BaseModel):
    trends: list[NewPatientsDataPoint]


# --- Top Medications ---
class TopMedicationItem(BaseModel):
    medication_name: str
    category: str
    count: int


class TopMedicationsResponse(BaseModel):
    items: list[TopMedicationItem]


# --- Recent Activity ---
class ActivityItem(BaseModel):
    event_type: str
    description: str
    actor_name: str
    patient_name: str
    patient_id: UUID
    timestamp: datetime.datetime
    detail: str | None = None


class RecentActivityResponse(BaseModel):
    items: list[ActivityItem]


# --- Pharmacy search ---
class PharmacyResult(BaseModel):
    name: str
    address: str
    lat: float
    lng: float
    distance: Optional[float] = None
    rating: Optional[float] = None
    open_now: Optional[bool] = None


class PharmacySearchResponse(BaseModel):
    results: list[PharmacyResult]
